[Docs] Add CodeArtifact user documentation (#9751)
mlrun/mlrun: docs/store/code-artifacts.md

(code-artifacts)=
# Code artifacts

> A code artifact (`kind="code"`) stores a function or workflow source file (or an archive of files) as a versioned MLRun artifact. Once logged, the artifact can be referenced by a `store://` URI as the source for {py:meth}`~mlrun.projects.MlrunProject.set_function` or {py:meth}`~mlrun.projects.MlrunProject.set_workflow`, instead of a local path or git source. The code is downloaded by the runner pod at runtime — the client never resolves the URI.

Reviewer: 1.  instead of a local path or git source => in addition to 
2.  remote url is also supported

Author: ok

**See also**
- {ref}`log-artifacts`
- {ref}`artifacts`
- {py:meth}`~mlrun.projects.MlrunProject.log_code_file`
- {py:meth}`~mlrun.projects.MlrunProject.set_function`
- {py:meth}`~mlrun.projects.MlrunProject.set_workflow`

## Log a code artifact

Use {py:meth}`~mlrun.projects.MlrunProject.log_code_file` to register a Python file or archive as a code artifact:

```python
# Minimal — local file uploaded to the project's artifact path
artifact = project.log_code_file("my_func_code", local_path="./my_func.py")
```

Common variants:

```python
# Upload to a specific remote target
project.log_code_file(
    "my_func_code",
    local_path="./my_func.py",
    target_path="s3://bucket/funcs/my_func.py",
)

# Reference an existing object via a datastore profile (recommended when
# credentials should be centralized server-side)
project.log_code_file(
    "my_func_code",
    target_path="ds://my-profile/funcs/my_func.py",
)

# Inline body (small files only)
project.log_code_file("my_func_code", body=b"def main():\n    return 'ok'\n")

# Archive containing multiple source files; extracted on resolution
project.log_code_file("my_pkg_code", local_path="./my_pkg.zip")
```

> `language` is auto-derived from the file suffix when omitted (`.py` / `.ipynb` → `"python"`). `code_type` defaults to `"function"`.

Reviewer: Besides describing `language` and `code_type` defaults, I think they should be used in the examples (`requirements` as well).
Update: I saw that you do have these described later - so either pull them earlier or reference their section.

Author: ok

## Run a function from a code artifact

Pass the artifact's `store://` URI as the `func` argument:

```python
fn = project.set_function(
    func="store://artifacts/my-project/my_func_code",
    name="my_func",
    kind="job",
    handler="my_func:main",
)

fn.run(params={"p1": 5})
```

> The `store://` URI is stored as-is on the function and in the MLRun DB. The runner pod resolves the artifact and downloads the code at startup — no client-side resolution.

Reviewer: ```suggestion
The `store://` URI is stored as-is in the function and in MLRun DB. The runner pod resolves the artifact and downloads the code at startup ( "pull at runtime")
```

Author: ok

> For Nuclio and serving functions, the default is **build-time** mode: the server downloads the code at deploy and embeds it in the processor image. Set `load_source_on_run=True` on the function spec to switch to **runtime** mode, where an init container fetches the code at pod startup (no image rebuild on code-only changes). See [function storage](../runtimes/function-storage.md) for the general source-loading mechanism.

Reviewer: ```suggestion
For Nuclio and serving functions, the default is "pull at buildtime" mode: the server downloads the code at deploy and embeds it in the processor image. Set `load_source_on_run=True` on the function spec to switch to "pull at runtime" mode, where an init container fetches the code at pod startup (no image rebuild on code-only changes). See [function storage](../runtimes/function-storage.md) for the general source-loading mechanism.
```

Author: ok

```python
artifact = project.log_code_file("my_pkg_code", local_path="./my_pkg.zip")

fn = project.set_function(
    func=artifact.uri,
    name="trainer",
    kind="job",
    handler="trainer:main",  # file inside the archive, not "my_pkg"
)
```

## Run a workflow from a code artifact

`set_workflow` accepts a `store://` URI as `workflow_path`:

```python
project.set_workflow(
    name="main",
    workflow_path="store://artifacts/my-project/my_workflow_code",
)

project.run(name="main", arguments={"p1": 5})
```

The workflow file is downloaded by the runner pod at execution time. Supported engines: `kfp`, `remote`, `remote:kfp`, `remote:local`, `local`.

## Requirements and rebuild behavior

A code artifact can declare its own pip dependencies:

```python
project.log_code_file(
    "my_func_code",
    local_path="./my_func.py",
    requirements=["pandas>=2.0", "numpy"],
)
```

At deploy or run time the server merges the artifact's `requirements` into `function.spec.build.requirements`. User-set requirements (via `func.with_requirements()`) take priority over artifact requirements, and deduplication is case-insensitive.

> Changing requirements always requires a re-deploy that rebuilds the image. Changing only the code has different costs:

Reviewer: > Changing requirements always requires a re-deploy that rebuilds the image

How to re-deploy rebuilding the image? please explain (for nuclio it's explained in the table and also before, but not for job).

Author: will do

# Inline body (small files only)
project.log_code_file("my_func_code", body=b"def main():\n    return 'ok'\n")

# Archive containing multiple source files; extracted on resolution
project.log_code_file("my_pkg_code", local_path="./my_pkg.zip")

# With an explicit code_type and pip dependencies
project.log_code_file(
    "my_workflow_code",
    local_path="./my_workflow.py",
    code_type="workflow",
    requirements=["pandas>=2.0"],
)
```

`language` is auto-derived from the file suffix when omitted (`.py` / `.ipynb` → `"python"`). `code_type` defaults to `"function"`. See [function vs workflow code_type](#function-vs-workflow-code_type) and [Requirements and rebuild behavior](#requirements-and-rebuild-behavior) for those parameters.

## Run a function from a code artifact

Pass the artifact's `store://` URI as the `func` argument:

```python
fn = project.set_function(
    func="store://artifacts/my-project/my_func_code",
    name="my_func",
    kind="job",
    handler="my_func:main",
)

fn.run(params={"p1": 5})
```

The `store://` URI is stored as-is in the function and in the MLRun DB. The runner pod resolves the artifact and downloads the code at startup ("pull at runtime").

> For Nuclio and serving functions, the default is "pull at buildtime" mode: the server downloads the code at deploy and embeds it in the processor image. Set `load_source_on_run=True` on the function spec to switch to "pull at runtime" mode, where an init container fetches the code at pod startup (no image rebuild on code-only changes). See [function storage](../runtimes/function-storage.md) for the general source-loading mechanism.

Reviewer: please move this into a subsection dealing with running functions of type nuclio/serving (and app-runtime - or give reference). should have description of the api needed (deploy etc) and then describe the diff behavior pull at buildtime/runtime

)
```

## Run a workflow from a code artifact

`set_workflow` accepts a `store://` URI as `workflow_path`:

```python
project.set_workflow(
    name="main",
    workflow_path="store://artifacts/my-project/my_workflow_code",
)

project.run(name="main", arguments={"p1": 5})
```

The workflow file is downloaded by the runner pod at execution time. Supported engines: `kfp`, `remote`, `remote:kfp`, `remote:local`, `local`.

## Requirements and rebuild behavior

A code artifact can declare its own pip dependencies:

```python
project.log_code_file(
    "my_func_code",
    local_path="./my_func.py",
    requirements=["pandas>=2.0", "numpy"],
)
```

At deploy or run time the server merges the artifact's `requirements` into `function.spec.build.requirements`. User-set requirements (via `func.with_requirements()`) take priority over artifact requirements, and deduplication is case-insensitive.

A change to the artifact's `requirements` is a build-configuration change, and like any other such change it is not detected automatically (`auto_build` does not pick it up). Rebuild the image as described in {ref}`build-function-image`.

> A change to the code only (no requirements change) depends on the source-loading mode described in [Run a function from a code artifact](#run-a-function-from-a-code-artifact). When the code is resolved at build time it is embedded in the image, so a code change is also a build-configuration change and the image must be rebuilt. When the code is resolved at runtime it is fetched by the pod at startup, so a code change is not a build-configuration change and no rebuild is needed. For job and KFP functions the next `run()` picks up the latest artifact version.

Reviewer: Very cumbersome explanation - and also deals with code-only change where we are in a section dealing with requirments.
